### PR TITLE
refactor: use `ScopedRef` (regions) in `MutMap`

### LINKED ISSUE #3306: Use `ScopedRef` in `MutMap`
Similar to #3300

### PR DESCRIPTION
**Pull request by jaschdoc**
I'm not sure where the error is currently. I've been developing this in a separate project using unique names s.t. it wouldn't clash with the standard library and that project does not compile yet (i.e. there's an error in the `MutMap` or its tests somewhere). [Additionally, the fixpoint interpreter (datalog related?) depends on `MutMap`](https://github.com/flix/flix/blob/2c03c87d7b35d676a720727978cb81aaadb07c96/main/src/library/Fixpoint/Interpreter.flix#L42).

### CODE REVIEW

**Comment by magnus-madsen** [comment]
OK, since the Datalog engine depends on this we should probably wait a bit.

**Comment by magnus-madsen** [comment]
Whats the error? Can you do it incrementally?

**Comment by jaschdoc** [comment]
No error yet. The compiler just crashes. Can give you a stack trace tomorrow :)

**Comment by jaschdoc** [comment]
It's probably some purity / region error I've made somewhere that I can't find.

**Comment by magnus-madsen** [comment]
I would kind-of need to know what function causes it.

**Comment by magnus-madsen** [comment]
It is "known" that the Boolean Unification sometimes chokes. We need an example to work on it :)

**Comment by jaschdoc** [comment]
> It is "known" that the Boolean Unification sometimes chokes. We need an example to work on it :)

I'll try to narrow it down.

**Comment by magnus-madsen** [comment]
It might also be useful to refactor the references tests and see where it starts to fail.

**Comment by jaschdoc** [comment]
Okay I think I've gained some insight into this.
The `merge` group of functions seem to cause some issues. First of all, the compiler times out (crashes?) in vscode even when both `m1` and `m2` share the same region (as shown below). However, when running tests the program compiles but tests fail.
```flix
    pub def merge!(m1: MutMap1[k, v, r], m2: MutMap1[k, v, r]): Unit  \ { Read(r), Write(r) } with Order[k] =
        mergeWithKey!((_, v1, _) -> v1, m1, m2)

    pub def mergeWith!(f: (v, v) -> v & ef, m1: MutMap1[k, v, r], m2: MutMap1[k, v, r]): Unit  \ { ef, Read(r), Write(r) } with Order[k] =
        mergeWithKey!((_, v1, v2) -> f(v1, v2), m1, m2)

    pub def mergeWithKey!(f: (k, v, v) -> v & ef, m1: MutMap1[k, v, r], m2: MutMap1[k, v, r]): Unit \ { ef, Read(r), Write(r) } with Order[k] =
        let MutMap1(mm1) = m1;
        let MutMap1(mm2) = m2;
        mm2 := Map.unionWithKey(f, deref mm1, deref mm2)
```
Example test
```flix
    @test
    def merge02(): Bool = region r {
        let m1 = new MutMap1(r);
        let m2 = new MutMap1(r);
        let _ = MutMap1.put!(1, m1);
        let _ = MutMap1.put!(2, m2);
        MutMap1.merge!(m1, m2);
        MutMap1.memberOf(1, m2) // and MutMap1.memberOf(2, m2)
    }
```

Finally, notice that I've written `let _ = MutMap1.put!(1, m1);` as the compiler throws the following error otherwise:
```
>> Useless expression: It has no side-effect(s) and its result is discarded.

96 |         MutMap1.put!(1, m1);
             ^^^^^^^^^^^^^^^^^^^
             useless expression.
```

Using latest compiler built from f8b8b76.

**Comment by magnus-madsen** [comment]
```
>> Useless expression: It has no side-effect(s) and its result is discarded.

96 |         MutMap1.put!(1, m1);
             ^^^^^^^^^^^^^^^^^^^
             useless expression.
```

Your put needs two arguments (key and value) otherwise it is a function waiting to be applied :)

**Comment by magnus-madsen** [comment]
Could you update the PR where you leave in the functions that seem to work and comment the other functions/tests.

**Comment by jaschdoc** [comment]
> While I get a complicated type for `mergeWithKey` I think it still goes through the compiler.

I fixed my test issues which fixed everything. However, when I decouple the regions for `m1` and `m2` the compiler times out and crashes yielding the stack trace from above.

> Could you update the PR where you leave in the functions that seem to work and comment the other functions/tests.

Sure thing :)

**Comment by jaschdoc** [comment]
Updated the PR with working functions

**Comment by magnus-madsen** [comment]
> Yes but we still have the Fixpoint interpreter dependency that needs to be refactored

which parts?

**Comment by magnus-madsen** [comment]
> 

I would like to wait with that. The region stuff should still work with impure.

**Comment by jaschdoc** [comment]
Okay I made something that compiles at least. The functions `{map, mapWithKey}` are broken until the boolean formulas are minimized--marking them as impure does not help.

**Comment by magnus-madsen** [comment]
Could you update #3306 with a list of the items that remain (and that have been temporary broken)?

Then we can possibly merge this in.

File: main/src/library/Map.flix [hunk 1]
```diff
@@ -752,13 +752,6 @@ namespace Map {
         let f = (acc, k, v) -> Map.insertWith(Set.union, v, Set#{k}, acc);
         Map.foldLeftWithKey(f, empty(), m)
 
-    ///
-    /// Returns `m` as a mutable set.
-    ///
-    @Time(1) @Space(1)
-    pub def toMutMap(m: Map[k, v]): MutMap[k, v] & Impure =
```
Comment (magnus-madsen) [inline main/src/library/Map.flix:759]: I think it should be possible to keep this, if it looks like:

pub def toMutMap(m: Map[k, v]): MutMap[k, v, Impure] & Impure = .... as MutMap[k, v, Impure]

does that not work?

Comment (magnus-madsen) [reply main/src/library/Map.flix:759]: Or something like this.

Comment (magnus-madsen) [reply main/src/library/Map.flix:759]: I think if we restore `toMutMap` then we should almost be ready to merge this in.

**Comment by magnus-madsen** [comment]
@jaschdoc Everything works now? No timeouts/crashes?

**Comment by magnus-madsen** [comment]
> Yes with a few exceptions: `map` and `mapWithKey` have been commented out since they crash the compiler. `{sameElements, merge!, mergeWith!, mergeWithKey!}` take maps `m1` and `m2` as parameters but currently they are required to share the same region. Uncoupling the regions s.t. `m1: MutMap[k, v, r1]` and `m2: MutMap[k, v, r2]` crashes the compiler.

Could we leave them in as impure? E.g. they would take an impure map and be impure?

I prefer that over commenting code.

**Comment by jaschdoc** [comment]
Okay should be good now

**Comment by magnus-madsen** [comment]
It works even without casting to impure?

**Comment by magnus-madsen** [comment]
So now no functions (other than empty?) are removed?

And the same for `MutSet`-- we did not delete any functions there either, right?

**Comment by jaschdoc** [comment]
`MutMap.empty` is still there to create a `MutMap` without a region (as far as I know the static keyword does not work yet i.e. `new MutMap(static)`)
I'll check again with `MutSet` but I'm fairly certain we did not delete anything there.

**Comment by magnus-madsen** [comment]
Roger. Thanks!

**Pull request merged**